### PR TITLE
Simplify Python modules installation

### PR DESCRIPTION
**Pull request by TimoWilken**
* Remove miniconda environment setup for MacOS -- tensorflow-metal can be installed with plain pip now.
* Deduplicate requirements specifications to reduce need for copy-pasting.
* Stop installing modules inside a fresh virtualenv, since those are not relocatable. Use `PYTHONPATH` instead.
* Remove deactivation of outer virtualenv, since that will mess up the PATH.

This works for me locally with the system Python and the alidist Python version on slc7 and slc8, with the standard setup on Mac (system Python) and Ubuntu (alidist Python).

@Barthelemy @awegrzyn: I wasn't sure what your intentions in overriding `PIP39_REQUIREMENTS` in `defaults-o2-dataflow.sh` were. As is, the assignment deleted all of the ML-related packages to be installed on that version, installing only the basics (e.g. ipython) and `setuptools==65.5.1`. I've deleted all Python packages to be installed and setting your preferred setuptools version in `PIP_BASE_REQUIREMENTS` instead -- please let me know if there's something I'm missing here!

Also Cc: @adriansev -- let me know if you see something wrong here!

### CODE REVIEW

**Comment by ktf** [comment]
Looks good to me. Please do not merge before we merge the C++20, to avoid invalidating the currently ongoing caching process.

**Comment by adriansev** [comment]
Hi @TimoWilken it looks _very_ good to me but i notices a small thing:
in python-modules there is `PYTHON_MODULES_INSTALLROOT=$INSTALLROOT/share/python-modules`
whereas in xjalienfs there is `--target="$INSTALLROOT/lib/python/site-packages"`
Moreover in python-modules there is `PYTHON_MODULES_INSTALLROOT=$INSTALLROOT/share/python-modules` and it has nicer she-bang fix than xjalienfs .. is there a possibility to migrate
these "goodies" also to xjalienfs :D ?
Thank you!

**Review by adriansev — APPROVED** [review]
LGTM

**Comment by TimoWilken** [comment]
Thanks @adriansev! I've changed the installation path to `.../lib/python/site-packages` to match the other things. Also the shebang patcher needed fixing (caught by the AliGenerators check).

I think the shebang patching in xjalienfs should stay as-is, since it handles a few more cases (including multiline polyglot shebangs).

@awegrzyn Do you see any issues with the installation path change for the RPMs that you build? Some Python-related packages (e.g. Python-modules and xjalienfs) will have some overlap between their `lib/python/site-packages` directories (e.g. both install `certifi`, `idna`, `requests`, etc).

For our updatable RPMs, this is not a problem, since we install to `/opt/alisw/<arch>/<package>`, so there's no overlap between packages.

**Comment by adriansev** [comment]
@TimoWilken Thanks a lot!! (for both work and explanation :) )

**Comment by awegrzyn** [comment]
@TimoWilken We also install RPMs in dedicated path (under `/opt/o2`)

**Comment by pzhristov** [comment]
It seems that with these changes Python is not compiled anymore but taken from the system. The result is that when we enter in the O2 environment, we cannot execute for example `ipython`. Here is the error I get:
'''
[O2/latest] ~/alisw $> ipython
Traceback (most recent call last):
  File "/Users/phristov/alisw/sw/osx_arm64/Python-modules/1.0-local1/bin/ipython", line 5, in <module>
    from IPython import start_ipython
  File "/Users/phristov/alisw/sw/osx_arm64/Python-modules/1.0-local1/lib/python/site-packages/IPython/__init__.py", line 55, in <module>
    from .terminal.embed import embed
  File "/Users/phristov/alisw/sw/osx_arm64/Python-modules/1.0-local1/lib/python/site-packages/IPython/terminal/embed.py", line 15, in <module>
    from IPython.core.interactiveshell import DummyMod, InteractiveShell
  File "/Users/phristov/alisw/sw/osx_arm64/Python-modules/1.0-local1/lib/python/site-packages/IPython/core/interactiveshell.py", line 339, in <module>
    class InteractiveShell(SingletonConfigurable):
  File "/Users/phristov/alisw/sw/osx_arm64/Python-modules/1.0-local1/lib/python/site-packages/IPython/core/interactiveshell.py", line 2908, in InteractiveShell
    @asyncio.coroutine
     ^^^^^^^^^^^^^^^^^
AttributeError: module 'asyncio' has no attribute 'coroutine'. Did you mean: 'coroutines'?
'''

**Comment by pzhristov** [comment]
@TimoWilken @ktf I think we should not drop the virtual environment where we run Python. This will help to avoid mismatches between the packages in the system installation and the customised list we install with alibuild.

**Comment by TimoWilken** [comment]
Hi @pzhristov, that looks like we just install an outdated version of `ipython` -- `asyncio.coroutine` was removed in Python 3.11. I've pushed a commit that updates the old ipython-related packages for Python 3.11+. That should fix it!

**Comment by pzhristov** [comment]
> Hi @pzhristov, that looks like we just install an outdated version of `ipython` -- `asyncio.coroutine` was removed in Python 3.11. I've pushed a commit that updates the old ipython-related packages for Python 3.11+. That should fix it!

Yes, I realised that it was because of obsolete packages. However, please restore the virtual environment to avoid issues with packages from the system or brew.

**Comment by TimoWilken** [comment]
Caching together with #5271: https://alijenkins.cern.ch/job/CacheO2Package/97/

**Comment by TimoWilken** [comment]
Caching succeeded and CI error is unrelated; merging.

**Comment by mfasDa** [comment]
I get a crash on macOS Sonoma with this error message:

```
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: # See version compatibility table at https://pypi.org/project/tensorflow-metal/
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: tensorflow-metal == 1.0.0; sys_platform == darwin and python_version >= 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2:
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: xgboost == 0.82; python_version < 3.8
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: xgboost == 1.3.3; python_version == 3.8
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: xgboost == 1.2.0; python_version >= 3.9 and python_version < 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: xgboost == 1.7.5; python_version >= 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2:
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: dryable == 1.0.3; python_version < 3.9
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: dryable == 1.0.5; python_version >= 3.9
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2:
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: pandas == 0.24.2; python_version < 3.8
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: pandas == 1.2.3; python_version == 3.8
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: pandas == 1.1.5; python_version >= 3.9 and python_version < 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: pandas == 1.5.3; python_version >= 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2:
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: dask[array,dataframe,distributed] == 2023.2.0; python_version < 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: dask[array,dataframe,distributed] == 2023.12.1; python_version >= 3.11
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: dask_jobqueue == 0.8.2
2024-02-09@10:28:18:DEBUG:QualityControl:Python-modules-list:aliceo2: ": File name too long
2024-02-09@10:28:18:ERROR:QualityControl:Python-modules-list:aliceo2: Error while executing /Users/markusfasel/alice/sw/SPECS/osx_arm64/Python-modules-list/1.0-local6/build.sh on `mfaselbook1.dyndns.cern.ch'.
2024-02-09@10:28:18:ERROR:QualityControl:Python-modules-list:aliceo2: Log can be found in /Users/markusfasel/alice/sw/BUILD/Python-modules-list-latest/log
2024-02-09@10:28:18:ERROR:QualityControl:Python-modules-list:aliceo2: Please upload it to CERNBox/Dropbox if you intend to request support.
2024-02-09@10:28:18:ERROR:QualityControl:Python-modules-list:aliceo2: Build directory is /Users/markusfasel/alice/sw/BUILD/Python-modules-list-latest/Python-modules-list.
```

**Comment by TimoWilken** [comment]
Hi @mfasDa! That looks like quoting gone wrong, but I can't see the full context. Could you attach a larger part of the output?

**Comment by TimoWilken** [comment]
Hi @mfasDa, I've tried to reproduce this on Sonoma, but I can't with aliBuild v1.16.1.

Do you have the latest aliBuild? (You can install it using `brew install alisw/system-deps/alibuild`.)

**Comment by mfasDa** [comment]
Hi @TimoWilken ,

Indeed I was still with aliBuild 1.15. After upgrading aliBuild from brew to 1.16.1 it works.

Cheers

Markus

**Pull request merged**
